### PR TITLE
Handle gracefully when an invalid devfile is found in a git repository

### PR DESCRIPTION
**Pull request by olexii4**
Signed-off-by: Oleksii Orel <oorel@redhat.com>

### What does this PR do?
Add a warning message for the user and prompt him if the workspace should start anyway (ignoring the existing devfile).

### What issues does this PR fix or reference?
fixes https://github.com/eclipse/che/issues/20738

### Is it tested? How?
1. Deploy Che (dashboard image: quay.io/eclipse/che-dashboard:pr-643).
2. Try to accept the factory with the wrong URL, like 
```$CHE_HOST#https://github.com/AObuchow/che-dashboardrrrr```.
An error message should be shown:
Знімок екрана 2022-10-06 о 17 08 30
3. Try to accept the factory from the repo with an invalid devfile scheme, like 
```$CHE_HOST#https://github.com/olexii4/che-dashboard```.
A warning message about starting anyway should be shown:
Знімок екрана 2022-10-14 о 17 54 00
4. Try to accept the factory from the repo with a valid devfile wich includes unsuported attributes for target platform, like 
```$CHE_HOST#https://github.com/AObuchow/che-dashboard```.
A warning message about starting anyway should be shown:
Знімок екрана 2022-10-14 о 17 53 41

### CODE REVIEW

File: packages/dashboard-frontend/src/containers/Loader/Factory/Steps/Apply/Devfile/index.tsx [hunk 1]
```diff
@@ -150,25 +173,45 @@ class StepApplyDevfile extends AbstractLoaderStep<Props, State> {
...
+          const sourceUrl = new URL(factoryParams.sourceUrl);
+          const name = getProjectName(factoryParams.sourceUrl);
+          const origin = sourceUrl.pathname.endsWith('.git')
+            ? `${sourceUrl.origin}${sourceUrl.pathname}`
+            : `${sourceUrl.origin}${sourceUrl.pathname}.git`;
+          _devfile.projects[0] = { git: { remotes: { origin } }, name };
+          // change default name
+          _devfile.metadata.name = name;
+          _devfile.metadata.generateName = name;
+        }
```
Comment (akurinnoy) [inline packages/dashboard-frontend/src/containers/Loader/Factory/Steps/Apply/Devfile/index.tsx:195]: Don't you mind moving these lines into the the `updateCurrentDevfile` method?

File: packages/dashboard-frontend/src/containers/Loader/Factory/Steps/Apply/Devfile/index.tsx [hunk 2]
```diff
@@ -206,7 +249,38 @@ class StepApplyDevfile extends AbstractLoaderStep<Props, State> {
     );
   }
 
+  private handleCreateWorkspaceError(): void {
+    const { defaultDevfile } = this.props;
+    const { devfile } = this.state;
+    const _devfile = defaultDevfile;
```
Comment (akurinnoy) [inline packages/dashboard-frontend/src/containers/Loader/Factory/Steps/Apply/Devfile/index.tsx:255]: You need to clone the default devfile, otherwise, it changes, isn't it?

```suggestion
    const _devfile = cloneDeep(defaultDevfile);
```

File: packages/dashboard-frontend/src/containers/Loader/Factory/Steps/Apply/Devfile/index.tsx [hunk 3]
```diff
@@ -126,12 +135,26 @@ class StepApplyDevfile extends AbstractLoaderStep<Props, State> {
     this.props.onRestart();
   }
 
+  private updateCurrentDevfile(devfile): void {
```
Comment (akurinnoy) [inline packages/dashboard-frontend/src/containers/Loader/Factory/Steps/Apply/Devfile/index.tsx:138]: Please the `devfile` parameter type.

File: packages/dashboard-frontend/src/store/DevfileRegistries/selectors.ts [hunk 4]
```diff
@@ -63,6 +67,31 @@ export const selectMetadataFiltered = createSelector(
...
+export const selectDefaultDevfile = createSelector(
+  selectState,
+  selectEmptyWorkspaceUrl,
+  (state, devfileLocation) => {
+    if (!devfileLocation) {
+      return undefined;
+    }
+    const devfileContent = state.devfiles[devfileLocation]?.content;
+    if (devfileContent) {
+      return load(devfileContent) as devfileApi.Devfile;
```
Comment (akurinnoy) [inline packages/dashboard-frontend/src/store/DevfileRegistries/selectors.ts:89]: I think you need to wrap this line with a try-catch.

File: packages/dashboard-frontend/src/containers/Loader/Factory/Steps/Apply/Devfile/index.tsx (hunk 2 above)
Comment (olexii4) [reply packages/dashboard-frontend/src/containers/Loader/Factory/Steps/Apply/Devfile/index.tsx:255]: It doesn't change because of this [line](https://github.com/eclipse-che/che-dashboard/pull/643/files#diff-a5725d88e5b0467a23748655268db89c00a2f5e7915a517a344f66a9b657bcb1R89):
```
return load(devfileContent) as devfileApi.Devfile;
```

Comment (akurinnoy) [reply packages/dashboard-frontend/src/containers/Loader/Factory/Steps/Apply/Devfile/index.tsx:255]: You're right. Thanks

File: packages/dashboard-frontend/src/containers/Loader/Factory/Steps/Apply/Devfile/index.tsx (hunk 1 above)
Comment (olexii4) [reply packages/dashboard-frontend/src/containers/Loader/Factory/Steps/Apply/Devfile/index.tsx]: Ok. I fixed it.

File: packages/dashboard-frontend/src/containers/Loader/Factory/Steps/Apply/Devfile/index.tsx (hunk 3 above)
Comment (olexii4) [reply packages/dashboard-frontend/src/containers/Loader/Factory/Steps/Apply/Devfile/index.tsx]: I fixed it.

File: packages/dashboard-frontend/src/store/DevfileRegistries/selectors.ts (hunk 4 above)
Comment (olexii4) [reply packages/dashboard-frontend/src/store/DevfileRegistries/selectors.ts]: Ok. I fixed it.

File: packages/dashboard-frontend/src/pages/Loader/Factory/CreateWorkspaceErrorItems/index.tsx [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+                  onToggle={isExpanded => this.onToggle(isExpanded)}
+                  direction="up"
+                >
+                  {expandableSectionTitle}
+                </ExpandableSectionToggle>
+              )}
+            </CodeBlock>
+          </Text>
+          <Text component={TextVariants.small}>
+            If you continue it will be ignored and a regular workspace will be created. You will You
```
Comment (ibuziuk) [inline packages/dashboard-frontend/src/pages/Loader/Factory/CreateWorkspaceErrorItems/index.tsx:157]:
```suggestion
            If you continue it will be ignored and a regular workspace will be created. You will
```

File: packages/dashboard-frontend/src/pages/Loader/Factory/__tests__/__snapshots__/applyingDevfileErrorItems.spec.tsx.snap [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+          </code>
+        </pre>
+      </div>
+    </div>
+  </small>
+  <small
+    className=""
+    data-pf-content={true}
+  >
+    If you continue it will be ignored and a regular workspace will be created. You will have have chance to fix the Devfile from the remote IDE once the dev environment is started.
```
Comment (ibuziuk) [inline packages/dashboard-frontend/src/pages/Loader/Factory/__tests__/__snapshots__/applyingDevfileErrorItems.spec.tsx.snap:95]:
```suggestion
    If you continue it will be ignored and a regular workspace will be created. You will have chance to fix the Devfile from the remote IDE once the dev environment is started.
```

File: packages/dashboard-frontend/src/pages/Loader/Factory/ApplyingDevfileErrorItems/index.tsx [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+                >
+                  {expandableSectionTitle}
+                </ExpandableSectionToggle>
+              )}
+            </CodeBlock>
+          </Text>
+          <Text component={TextVariants.small}>
+            If you continue it will be ignored and a regular workspace will be created. You will
+            have have chance to fix the Devfile from the remote IDE once the dev environment is
+            started.
```
Comment (ibuziuk) [inline packages/dashboard-frontend/src/pages/Loader/Factory/ApplyingDevfileErrorItems/index.tsx:157]:
```suggestion
```

File: packages/dashboard-frontend/src/store/FactoryResolver/index.ts [hunk 8]
```diff
@@ -212,8 +212,7 @@ export const actionCreators: ActionCreators = {
             throw response.data;
           }
         }
-        const errorMessage =
-          'Failed to request factory resolver: ' + common.helpers.errors.getMessage(e);
```
Comment (akurinnoy) [inline packages/dashboard-frontend/src/store/FactoryResolver/index.ts:216]: I was just wondering why did you change this error message?

File: packages/dashboard-frontend/src/containers/Loader/Factory/Steps/Apply/Devfile/index.tsx [hunk 9]
```diff
@@ -126,12 +136,45 @@ class StepApplyDevfile extends AbstractLoaderStep<Props, State> {
     this.props.onRestart();
   }
 
+  private updateCurrentDevfile(devfile: devfileApi.Devfile): void {
+    const { factoryResolver, allWorkspaces, defaultDevfile } = this.props;
+    const { factoryParams } = this.state;
+    const { factoryId, policiesCreate, storageType } = factoryParams;
+    // check if it is a using default devfile flow
```
[the remark this answers is not in the document]
Comment (olexii4) [reply packages/dashboard-frontend/src/containers/Loader/Factory/Steps/Apply/Devfile/index.tsx]: Ok. I have fixed it.

File: packages/dashboard-frontend/src/pages/Loader/Factory/ApplyingDevfileErrorItems/index.tsx [hunk 10]
```diff
@@ -0,0 +1,164 @@
...
+      isExpanded: false,
+      errorMessage,
+    };
+
+    this.checkOverflow = () => {
+      if (this.state.isExpanded) {
+        this.setState({
+          isExpanded: false,
+        });
+      }
```
[the remark this answers is not in the document]
Comment (olexii4) [reply packages/dashboard-frontend/src/pages/Loader/Factory/ApplyingDevfileErrorItems/index.tsx]: it is needed for analyze if it has an overflow after scaling

File: packages/dashboard-frontend/src/pages/Loader/Factory/ApplyingDevfileErrorItems/index.tsx [hunk 11]
```diff
@@ -0,0 +1,164 @@
...
+        const { offsetWidth, scrollWidth } = errorMessageElement;
+        if (scrollWidth > offsetWidth) {
+          this.setState({
+            hasExpand: true,
+          });
+        } else {
+          this.setState({
+            hasExpand: false,
+          });
+        }
```
[the remark this answers is not in the document]
Comment (olexii4) [reply packages/dashboard-frontend/src/pages/Loader/Factory/ApplyingDevfileErrorItems/index.tsx]: Ok. I have fixed it.

File: packages/dashboard-frontend/src/pages/Loader/Factory/ApplyingDevfileErrorItems/index.module.css [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+
+.error {
+  border-top: 2px solid;
+  border-color: #c9190b;
+  background-color: #faeae8;
```
[the remark this answers is not in the document]
Comment (olexii4) [reply packages/dashboard-frontend/src/pages/Loader/Factory/ApplyingDevfileErrorItems/index.module.css]: Ok. I have fixed it.

File: packages/dashboard-frontend/src/pages/Loader/Factory/ApplyingDevfileErrorItems/index.module.css [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+
+.error .textMessage {
+  font-size: small;
+  overflow: hidden;
+  text-overflow: ellipsis;
+}
+
+.error button {
+  font-size: small;
+}
```
[the remark this answers is not in the document]
Comment (olexii4) [reply packages/dashboard-frontend/src/pages/Loader/Factory/ApplyingDevfileErrorItems/index.module.css]: Ok. I have fixed it.

File: packages/dashboard-frontend/src/pages/Loader/Factory/ApplyingDevfileErrorItems/index.module.css [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+ */
+
+.error {
+  border-top: 2px solid;
+  border-color: #c9190b;
+  background-color: #faeae8;
+}
+
+.error .textMessage {
+  font-size: small;
```
[the remark this answers is not in the document]
Comment (olexii4) [reply packages/dashboard-frontend/src/pages/Loader/Factory/ApplyingDevfileErrorItems/index.module.css]: Ok. I have fixed it.

File: packages/dashboard-frontend/src/store/FactoryResolver/index.ts (hunk 8 above)
Comment (olexii4) [reply packages/dashboard-frontend/src/store/FactoryResolver/index.ts:216]: I removed this additional because the original error message starts from the same information.

**Review by akurinnoy — APPROVED** [review]
